types/src/v0/impls/block/full_payload/ns_proof/avidm.rs

```diff
@@ -0,0 +1,180 @@
+//! This module contains the namespace proof implementation for the new AvidM scheme.
+
+use hotshot_types::{data::VidCommitment, vid::avidm::AvidMCommon};
+use vid::avid_m::namespaced::NsAvidMScheme;
+
+use crate::{v0_3::AvidMNsProof, NamespaceId, NsIndex, NsPayload, NsTable, Payload, Transaction};
+
+impl AvidMNsProof {
+    pub fn new(payload: &Payload, index: &NsIndex, common: &AvidMCommon) -> Option<AvidMNsProof> {
+        let payload_byte_len = payload.byte_len();
+        let index = index.0;
+        let ns_table = payload.ns_table();
+        let ns_table = ns_table
+            .iter()
+            .map(|index| ns_table.ns_range(&index, &payload_byte_len).0)
+            .collect::<Vec<_>>();
+
+        if index >= ns_table.len() {
+            tracing::warn!("ns_index {:?} out of bounds", index);
+            return None; // error: index out of bounds
+        }
+
+        if ns_table[index].is_empty() {
+            None
+        } else {
+            match NsAvidMScheme::namespace_proof(common, &payload.raw_payload, index, ns_table) {
+                Ok(proof) => Some(AvidMNsProof(proof)),
+                Err(e) => {
+                    tracing::error!("error generating namespace proof: {:?}", e);
+                    None
+                },
+            }
+        }
+    }
+
+    /// Unlike the ADVZ scheme, this function won't fail with a wrong `ns_table`.
+    /// It only uses `ns_table` to get the namespace id.
+    pub fn verify(
+        &self,
+        ns_table: &NsTable,
+        commit: &VidCommitment,
+        common: &AvidMCommon,
+    ) -> Option<(Vec<Transaction>, NamespaceId)> {
+        match commit {
+            VidCommitment::V1(commit) => {
+                match NsAvidMScheme::verify_namespace_proof(common, commit, &self.0) {
+                    Ok(Ok(_)) => {
+                        let ns_id = ns_table.read_ns_id(&NsIndex(self.0.ns_index))?;
+                        let ns_payload = NsPayload::from_bytes_slice(&self.0.ns_payload);
+                        Some((ns_payload.export_all_txs(&ns_id), ns_id))
+                    },
+                    Ok(Err(_)) => None,
+                    Err(e) => {
+                        tracing::warn!("error verifying namespace proof: {:?}", e);
+                        None
+                    },
+                }
+            },
+            _ => None,
```

Comment:
why don't we verify V2?

Reply:
We don't have V2. V0 is handled in the advz module.